docs/src/neptune-core/overview.md

```diff
@@ -67,25 +66,38 @@ Different tasks are scheduled in the main loop every N seconds. These currently
 
 ## Central Primitives
 From `tokio`
 - `spawn`
 - `select!`
-- `tokio::sync::Mutex`
+- `tokio::sync::RwLock`
 
 From Std lib:
 - `Arc`
-- `std::sync::Mutex`
+
+From neptune-core:
+- `neptune_core::locks::tokio::AtomicRw`  (wraps `Arc<tokio::sync::RwLock>`)
 
 ## Persistent Memory
-We use `rusty-leveldb` for our database layer with a custom-wrapper that makes it more type safe. `rusty-leveldb` allows for atomic writes within *one* database which is equivalent to a table in SQL lingo. So if you want atomic writes across multiple datatypes (you do want this!) you need to put that `enum` into the database and then cast the output type to the correct type. I think this is a low price to pay to achieve atomicity on the DB-layer.
 
-Blocks are stored on disk and their position on disk is stored in the `block_index` database. Blocks are read from and written to disk using `mmap`.
+We use `leveldb` for our database layer with custom wrappers that make it more async-friendly, type safe, and emulate multi-table transactions.
+
+`neptune_core::database::NeptuneLevelDb` provides async wrappers for leveldb APIs to avoid blocking async tasks.
+
+`leveldb` is a simple key/value store, meaning it only allows manipulating individual strings. It does however provide a batch update facility. `neptune_core::database::storage::storage_schema::DbSchema` leverages these batch updates to provide vector and singleton types that can be manipulated in rust code and then atomically written to `leveldb` as a single batch update (aka transaction).
+
+Blocks are stored on disk and their position on disk is stored in the `block_index` database. Blocks are read from and written to disk using `mmap`.  We wrap all file-system calls with tokio's `spawn_blocking()` so they will not block other async tasks.
 
 ## Challenges
-- Deadlocks. Solution: always acquire locks in the same order. Note though that locks from `std::sync` may not be held over an `await`. The linter should tell you if you do this. When a function requires more than one lock, **the only correct ordering in which to acquire these locks is the order in which the fields are defined in `GlobalState`. Any deviation from this is a bug.**
+
+- Deadlocks. We only have a single RwLock over the GlobalState. This is encapsulated in struct `GlobalStateLock`. This makes deadlocks pretty easy to avoid, following some simple rules:
+
+  1. avoid deadlocking yourself. If a function has read-acquired the global lock then it must be released before write-acquiring. Likewise never attempt to write-acquire the lock twice.
+
+  2. avoid deadlocking others. Always be certain that the global lock will be released in timely fashion. In other words if you have some kind of long running task with an event loop that needs to acquire the global lock, ensure that it gets acquired+released inside the loop rather than outside.
```

Comment:
well it could just be temporarily blocking if the lock eventually gets released.  But I was trying to warn against a task with a "forever loop" that acquires the lock before the loop, and thus never releases it, and effectively deadlocks the program should any other task ever try to acquire it.

also, I like the symmetry of 1. yourself, and 2. others.

So I'm inclined to leave as-is, unless you have a suggestion how to make it clearer?